### PR TITLE
Pass generate_decimal256_case integration test, add `DataType::Decimal256`

### LINKED ISSUE #2093: Pass generate_decimal256_case integration test
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Decimal256Array should pass C++ integration test.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2093.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/datatypes/datatype.rs [hunk 1]
```diff
@@ -195,6 +195,8 @@ pub enum DataType {
     ///
     /// For example the number 123.45 has precision 5 and scale 2.
     Decimal(usize, usize),
+    /// Exact decimal value with 256 bits width
+    Decimal256(usize, usize),
```
Comment (alamb) [inline arrow/src/datatypes/datatype.rs:199]: Should we also rename `Decimal` to `Decimal128`? 

Or maybe use an extra mode in Decimal like `

```rust
enum BitWidth {
  // 128 bits
  Bits128,
  // 256 bits
  Bits256,
}


enum DataType {
..
  Decimal(usize, usize, DecimalWidth)
}

```

**Review by alamb — COMMENTED** [review]
This is very cool @viirya 

I wonder if anyone else has thoughts about how to name the new DataType?

File: arrow/src/datatypes/datatype.rs (hunk 1 above)
Comment (viirya) [reply arrow/src/datatypes/datatype.rs:199]: I thought to add a parameter to current `Decimal` like you put above. But soon I realize that I need to modify many places. And we need to do bit width check there. As in C++ and Python implementations, they have Decimal128Type and Decimal256Type separately. I feel that it is simple to work with.

And, yea, we need to rename `Decimal` to `Decimal128`. Although I've not done it in this change as this is basically for Decimal256 interop test. I may work on it in another PR.

**Comment by alamb** [comment]
I see https://github.com/apache/arrow-rs/issues/2101 as well so 👍

**Review by alamb — APPROVED** [review]

File: arrow/src/datatypes/datatype.rs (hunk 1 above)
Comment (alamb) [reply arrow/src/datatypes/datatype.rs:199]: > And, yea, we need to rename Decimal to Decimal128. 

I agree

>  I may work on it in another PR.

Thank you!

**Pull request merged**